### PR TITLE
refactor: integrated base Observer structure into existing EVM/BTC observers

### PR DESCRIPTION
**Pull request by ws4charlie**
# Description

The PR integrated base `Observer` structures into existing EVM/BTC `Signer` structures.

Closes: [2022](https://github.com/zeta-chain/node/issues/2022)
Closes: [2126](https://github.com/zeta-chain/node/issues/2126)

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Include instructions and any relevant details so others can reproduce. 

- [x] Tested CCTX in localnet
- [ ] Tested in development environment
- [x] Go unit tests
- [ ] Go integration tests
- [ ] Tested via GitHub Actions 

# Checklist:

- [ ] I have added unit tests that prove my fix feature works

## Summary by CodeRabbit

- **New Features**
  - Integrated base Observer structure into EVM/Bitcoin Observer.
- **Bug Fixes**
  - Simplified initialization of observers.
  - Enhanced error handling during observer creation.
- **Tests**
  - Added tests for opening/closing the database, saving, and reading/writing the last block scanned.
- **Refactor**
  - Updated observer method calls and property handling for consistency.
- **Documentation**
  - Updated changelog to reflect integration changes in observers.

### CODE REVIEW

File: zetaclient/chains/bitcoin/observer/observer.go [hunk 1]
```diff
@@ -125,157 +102,120 @@ type Observer struct {
...
+		pendingNonce:      0,
+		utxos:             []btcjson.ListUnspentResult{},
+		includedTxHashes:  make(map[string]bool),
+		includedTxResults: make(map[string]*btcjson.GetTransactionResult),
+		broadcastedTx:     make(map[string]string),
+		logger: Logger{
+			ObserverLogger: *baseObserver.Logger(),
+			UTXOs:          baseObserver.Logger().Chain.With().Str("module", "utxos").Logger(),
+		},
+		Mu: &sync.Mutex{},
```
[the remark this answers is not in the document]
Comment (ws4charlie) [reply zetaclient/chains/bitcoin/observer/observer.go]: Mutex had already been moved to base struct

File: zetaclient/chains/bitcoin/observer/observer.go [hunk 2]
```diff
@@ -609,158 +523,115 @@ func (ob *Observer) SaveBroadcastedTx(txHash string, nonce uint64) {
...
-// GetTxResultByHash gets the transaction result by hash
-func GetTxResultByHash(
-	rpcClient interfaces.BTCRPCClient,
-	txID string,
-) (*chainhash.Hash, *btcjson.GetTransactionResult, error) {
-	hash, err := chainhash.NewHashFromStr(txID)
+// GetBlockByNumberCached gets cached block (and header) by block number
+func (ob *Observer) GetBlockByNumberCached(blockNumber int64) (*BTCBlockNHeader, error) {
+	if result, ok := ob.BlockCache().Get(blockNumber); ok {
+		return result.(*BTCBlockNHeader), nil
```
[the remark this answers is not in the document]
Comment (ws4charlie) [reply zetaclient/chains/bitcoin/observer/observer.go]: https://github.com/zeta-chain/node/pull/2359/commits/642e54e5838a5c18a449342cdb6788a546d6a5d0

File: cmd/zetaclientd/utils.go [hunk 3]
```diff
@@ -116,33 +120,75 @@ func CreateChainObserverMap(
...
+
+		// create EVM client
+		evmClient, err := ethclient.Dial(evmConfig.Endpoint)
+		if err != nil {
+			logger.Std.Error().Err(err).Msgf("error dailing endpoint %s", evmConfig.Endpoint)
+			continue
+		}
+
+		// create EVM chain observer
+		co, err := evmobserver.NewObserver(
```
[the remark this answers is not in the document]
Comment (ws4charlie) [reply cmd/zetaclientd/utils.go]: https://github.com/zeta-chain/node/pull/2359/commits/642e54e5838a5c18a449342cdb6788a546d6a5d0

**Review by lumtis — APPROVED** [review]
Would be good to walk through CodeRabbit comments. Seems like some are pertinent

**Comment by ws4charlie** [comment]
> Would be good to walk through CodeRabbit comments. Seems like some are pertinent
resolved a typo in change log and log print.

**Comment by lumtis** [comment]
@ws4charlie can we resolve code rabbit non relevant review comments?

File: zetaclient/chains/base/observer.go [hunk 4]
```diff
@@ -232,9 +250,20 @@ func (ob *Observer) DB() *gorm.DB {
 	return ob.db
 }
 
+// WithTelemetryServer attaches a new telemetry server to the observer.
+func (ob *Observer) WithTelemetryServer(ts *metrics.TelemetryServer) *Observer {
```
[the remark this answers is not in the document]
Comment (swift1337) [reply zetaclient/chains/base/observer.go:254]: I also think that we need to use dependency injection in chain-specific observers i.e. they should accept `*base.Observer` as a first argument (e.g.  `evm.NewObserver(baseObserver, ...)`)

File: zetaclient/chains/base/observer.go [hunk 5]
```diff
@@ -28,9 +30,12 @@ const (
...
 	DefaultBlockCacheSize = 1000
 
-	// DefaultHeadersCacheSize is the default number of headers that the observer will keep in cache for performance (without RPC calls)
+	// DefaultHeaderCacheSize is the default number of headers that the observer will keep in cache for performance (without RPC calls)
 	// Cached headers can be used to get header information
-	DefaultHeadersCacheSize = 1000
+	DefaultHeaderCacheSize = 1000
+
+	// TempSQLiteDBPath is the temporary in-memory SQLite database used for testing
+	TempSQLiteDBPath = "file::memory:?cache=shared"
```
[the remark this answers is not in the document]
Comment (ws4charlie) [reply zetaclient/chains/base/observer.go:38]: Compared to creating a real folder, running unit tests (non-database-related) is faster on in-memory db.

[the remark this answers is not in the document]
Comment (0xBuzzLightyear) [reply zetaclient/chains/base/observer.go:38]: 100%, I meant that let's keep this in test suite, not `observer.go`

File: zetaclient/chains/base/observer.go [hunk 6]
```diff
@@ -72,6 +77,10 @@ type Observer struct {
 	// logger contains the loggers used by observer
 	logger ObserverLogger
 
+	// mu protects fields from concurrent access
+	// Note: base observer simply provides the mutex. It's the sub-struct's responsibility to use it to be thread-safe
+	mu *sync.Mutex
```
[the remark this answers is not in the document]
Comment (ws4charlie) [reply zetaclient/chains/base/observer.go:82]: using `mu sync.Mutex` cause compile errors when creating the concrete Observer/Signer. For instance, the `NewObserver` will copy the base struct's `sync.Mutex` into itself and compiler doesn't allow it.

File: zetaclient/chains/base/observer.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (ws4charlie) [reply zetaclient/chains/base/observer.go:254]: A few of the fields needs to be updated at runtime. Most of them (the others) are currently used by unit tests to mock up data.

File: zetaclient/chains/base/observer.go [hunk 7]
```diff
@@ -251,45 +280,69 @@ func (ob *Observer) WithLogger(logger Logger) *Observer {
 	return ob
 }
 
-// Stop returns the stop channel for the observer.
-func (ob *Observer) Stop() chan struct{} {
+// Mu returns the mutex for the observer.
+func (ob *Observer) Mu() *sync.Mutex {
```
[the remark this answers is not in the document]
Comment (ws4charlie) [reply zetaclient/chains/base/observer.go:282]: Same compiler errors.
`copylocks: literal copies lock value from *baseSigner: github.com/zeta-chain/zetacore/zetaclient/chains/base.Signer`

File: zetaclient/chains/base/observer.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (swift1337) [reply zetaclient/chains/base/observer.go:80]: https://go.dev/play/p/FbI5gba_aUb

File: zetaclient/chains/base/observer.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (ws4charlie) [reply zetaclient/chains/base/observer.go]: okay.

File: zetaclient/chains/base/observer.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (ws4charlie) [reply zetaclient/chains/base/observer.go:80]: A pointer of base works for me.

File: zetaclient/chains/base/observer.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (ws4charlie) [reply zetaclient/chains/base/observer.go:282]: https://github.com/zeta-chain/node/pull/2359/commits/ab0c960e115e1fa5f29b93d7bc2e3e80bc7e1429

File: zetaclient/chains/base/observer.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (swift1337) [reply zetaclient/chains/base/observer.go:252]: Okay, let's address those in the following PRs 👌

**Review by swift1337 — APPROVED** [review]
LGTM

**Pull request merged**